src/CodeMirrorBlocks.tsx

```diff
@@ -0,0 +1,69 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import ToggleEditor from './ui/ToggleEditor';
+import type {ToggleEditorProps} from './ui/ToggleEditor';
+import Args from './components/Args';
+import {DropTarget} from './components/DropTarget';
+import Node from './components/Node';
+import {AST} from './ast';
+import * as Nodes from './nodes';
+import * as NodeSpec from './nodeSpec';
+import * as Languages from './languages';
+import Pretty from 'pretty-fast-pretty-printer';
+import { PrimitiveGroup } from './parsers/primitives';
+
+type CodeMirrorBlocksOptions = {
+  value?: string;
+};
+
+// Consumes a DOM node to host the editor, a language object and the code
+// to render. Produces an object-representation of CMB, allowing for
+// integration with external (non-react) code
+export default class CodeMirrorBlocks {
+  constructor(container: Element, options: CodeMirrorBlocksOptions = {}, language: ToggleEditorProps['language'], cmOptions = {}) {
+    let api = {};
+    let initialCode = options.value;
+    ReactDOM.render(
+      <ToggleEditor
+        language={language}
+        initialCode={(initialCode == null) ? "" : initialCode}
+        api={api}
+        appElement={container}
+        options={options}
+        cmOptions={cmOptions}
+      />,
+      container
+    );
+    Object.assign(api, this.buildAPI());
+    // TODO(pcardune): What the heck is even going on here?
+    // constructors shouldn't return things that are not instances
+    // of the class they are constructing. We should not have to
+    // cast the return value to any for typescript to be happy.
+    return api as any;
```

Comment:
We need the CMB constructor to produce an object that behaves the same way as the result of a CodeMirror constructor. I couldn't figure out how to replicate this behavior _without_ producing an objection with functions that close over the class, but if you have a more elegant solution I would love to see it!

Reply:
oh, I think I see what's going on. Rendering a `ToggleEditor` component actually has side effects that cause `api` to be populated with stuff, which you then augment further with `buildAPI`. And you're saying you really need to be able to use the `new` keyword to create a new instance instead of just calling a function that returns an object?

Reply:
This is going back a bit, but if IIRC that wasn't the only issue. I think I kept getting the _ReactElement_ object back, not the instance itself.

I would happily take a PR if the only external difference is dropping the `new` keyword.

Reply:
Looks like it started here https://github.com/bootstrapworld/codemirror-blocks/commit/c9e1f7ce58ead23631700484d92e89dcb0572395#diff-53a8436ea52bed1ccdc7170a35a53faefa7e5cfcd605118132e0d8cbe31c4328R24-R33 and then was made further interesting with this commit: https://github.com/bootstrapworld/codemirror-blocks/commit/ea5694893735030c9ea1b09f04941977228aedbb#

It probably stemmed from before the CodeMirrorBlocks class existed and we were just calling ReactDOM.render(...).

The more I read through the code, the more I see that monkey-patching is used as the primary method for exposing an external API. This makes the type declarations really interesting...

The problem with dropping the `new` keyword is that it would be a breaking API change. What is your process for rolling out breaking changes?

Reply:
Since only WeScheme and Pyret are consumers of the library (for now), we don't have a process other than me committing to WeScheme and then poking Joe for his work on CPO. 

Totally safe to make the change.

Reply:
ok, that's great. I'll plan on submitting a separate PR for that once this PR gets merged.

Reply:
So I just tried this out to refresh my memory on the semantics of functions vs classes and we can actually just convert this to a function without it being a breaking change as the `new` keyword will still work. Class constructors are just functions that force you to use `new`:

```js
class Foo {
  constructor() {
    return {foo: "This is not an instance of Foo"};
  }
}
function Bar() {
  return {bar: "This is not an instance of Bar"};
}

// using the new keyword:
let foo = new Foo()
let bar = new Bar()
foo instanceof Foo // false! because of course it isn't...
bar instanceof Bar // also false
foo.foo // "This is not an instance of Foo"
bar.bar // "This is not an instance of Bar"

// not using the new keyword
foo = Foo() // TypeError: Class constructor Foo cannot be invoked without 'new'
bar = Bar() // works just fine.
bar.bar // "This is not an instance of Bar"
```

Of course typescript won't allow you to use `new` with a function that's not meant to be a constructor, but that's not a breaking change since no types were previously being exported with the CodeMirrorBlocks package.